sebs/gcp/datastore.py

```diff
@@ -0,0 +1,180 @@
+            return None
+
+        return table
+
+    def retrieve_cache(self, benchmark: str) -> bool:
+
+        if benchmark in self._benchmark_resources:
+            return True
+
+        cached_storage = self.cache_client.get_nosql_config(self.deployment_name(), benchmark)
+        if cached_storage is not None:
+            self._benchmark_resources[benchmark] = BenchmarkResources.deserialize(cached_storage)
+            return True
+
+        return False
+
+    def update_cache(self, benchmark: str):
+
+        self._cache_client.update_nosql(
+            self.deployment_name(), benchmark, self._benchmark_resources[benchmark].serialize()
+        )
+
+    def benchmark_database(self, benchmark: str) -> str:
+        return self._benchmark_resources[benchmark].database
+
+    def writer_func(
+        self,
+        benchmark: str,
+        table: str,
+        data: dict,
+        primary_key: Tuple[str, str],
+        secondary_key: Optional[Tuple[str, str]] = None,
+    ):
+
+        res = self._benchmark_resources[benchmark]
+        table_name = self._get_table_name(benchmark, table)
+
+        # FIXME: support both options
+        assert secondary_key is not None
+
+        if res.database_client is None:
+            res.database_client = datastore.Client(database=res.database)
+
+        parent_key = res.database_client.key(secondary_key[0], secondary_key[1])
+        key = res.database_client.key(
+            # kind determines the table
+            table_name,
+            # main ID key
+            secondary_key[1],
+            # organization key
+            parent=parent_key,
+        )
+
+        val = datastore.Entity(key=key)
+        val.update(data)
+        res.database_client.put(val)
+
+    def create_table(
+        self, benchmark: str, name: str, primary_key: str, _: Optional[str] = None
+    ) -> str:
+
+        benchmark_resources = self._benchmark_resources.get(benchmark, None)
+
+        if benchmark_resources is not None and name in benchmark_resources.kinds:
+            self.logging.info(f"Using cached Datastore kind {name}")
+            return name
+
+        """
+            No data for this benchmark -> we need to allocate a new Datastore database.
+        """
+
+        if benchmark_resources is None:
+
+            database_name = f"sebs-benchmarks-{self._cloud_resources.resources_id}-{benchmark}"
+
+            try:
+
+                self._cli_instance.execute(
+                    "gcloud firestore databases describe "
+                    f" --database='{database_name}' "
+                    " --format='json'"
+                )
+
+            except RuntimeError as e:
+
+                if "NOT_FOUND" in str(e):
+
+                    """
+                    Allocate a new Firestore database, in datastore mode
+                    """
+
+                    self.logging.info(f"Allocating a new Firestore database {database_name}")
+                    self._cli_instance.execute(
```

Comment:
What's the reason why we are using the CLI to do this?

Reply:
@oanarosca AFAIK there's no API way to allocate a Firestore database except of using `gcloud` tool. Did you find it somewhere? It would greatly simplify the implementation :-)